File: solutions/typescript/0020.ts
function isValid(s: string): boolean {
    const stack: string[] = [];
    const map: { [key: string]: string } = {
        '(': ')',
        '{': '}',
        '[': ']'
    };
    for (const char of s) {
        if (map[char]) {
            stack.push(char);
        } else if (stack.length > 0 && map[stack[stack.length - 1]] === char) {
            stack.pop();
        } else {
            return false;
        }
    }
    return stack.length === 0;
}

File: solutions/python/0020.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        map = {
            '(': ')',
            '{': '}',
            '[': ']'
        }
        for char in s:
            if char in map:
                stack.append(char)
            else:
                if not stack or map[stack.pop()] != char:
                    return False
        return not stack